### ISSUE TITLE
Correct minus sign is not interpreted as minus

### ISSUE THREAD

**Issue by phun-ky** [tc39/proposals#37]
The minus sign, `−` (U+2212) is not correctly recognized in JavaScript as minus:

    10 − 4
    // Yields: SyntaxError: illegal character
    10 - 4
    // Yields: 6

Is this an issue? Depends. Technically?  No, `−` is not in the standard keyboard layout, oddly. The issue comes to screen readers, i.e. readers used by blind or visually impaired programmers that code. Screen readers will read the normal `-` in `-1000` as: "Dash thousand", but if `−` is used, the screen reader will read: "Minus thousand". 

Another example is:

    1000 − 100
    // Screen reader: "Thousand minus hundred"
    1000 - 100
    // Screen reader: "Thousand dash hundred"

I think I need some more detail to back this up, but this is at my thoughts on this subject.

I am a fan of correctness and I would like to see JavaScript recognizing the correct symbols :)

**Comment by arcanis** [comment]
Just by curiosity, what's the position of others programming languages about this? I definitely see the reasoning behind this, but aren't you worried that it could eventually lead to source codes that use both the minus sign *and* the dash sign? Wouldn't that make it significantly worse for screen reading, since you would have two different vocals for a single expression? It would significantly complexify grepping such expressions.

**Comment by xtuc** [comment]
> is not in the standard keyboard layout

Would also be difficult to write.

**Comment by ljharb** [comment]
Please see https://github.com/tc39/proposals/blob/master/CONTRIBUTING.md for ways to contribute to ECMAScript - I'd suggest emailing esdiscuss.

**Comment by phun-ky** [comment]
@ljharb thanks for the information on that! :)

@xtuc 
> I think most parser will consider it as a non valid character.
Well, if you want to generalize JavaScript as just another language, sure. What another parser does or does is irrelevant tbh. 

This is not a technical issue, but a usability issue.  JavaScript should be in the forefront of making not only apps accessible, but also coding with it accessible.  I know that this is something that seem irrelevant, but I think we have to look at the bigger picture here. Thanks for the feedback and patience, will do some research on this before I send an email to esdicuss :)

**Comment by mk-pmb** [comment]
Very nice idea! However I doubt it will ever go in the spec, as parser people really love 7-bit ASCII. =)

**Comment by ljharb** [comment]
Alternatively you could follow the overwhelmingly longstanding best practice of using a linter, which with editor integration would immediately inform you of the parsing error.

**Comment by mk-pmb** [comment]
Oh right, for voice input you'd need some adjustments as well. However I doubt we're anywhere near a practical and convenient approach for dictating source code.

**Comment by phun-ky** [comment]
@mk-pmb thanks for the feedback, but here's a thought, what happens when math (like MathML) with correct symbols are introduced in the code? :fire:

**Comment by mk-pmb** [comment]
Since MathML can appear in other contexts than JS, there should be separate, context-independent MathML accessibility modules. JS source code accessibility modules should aim to integrate those.

**Comment by phun-ky** [comment]
After some hiatus on this issue, I am ready to revisit this. Just to be clear about parser vs screen reader responsibility areas; A screen reader, or a plugin, can never 100% correctly identify when to interpret a dash as a minus, but a parser can. 

I will try to summarize my thoughts and email the discussion group <3

**Comment by netizen-ais** [comment]
Just 2¢: Besides the security implications that could come with unicode parsing/tokenization, the [unicode table](https://www.unicode.org/charts/PDF/U0000.pdf) correctly defines the keyboard hypen (-) as HYPHEN-MINUS

**Comment by phun-ky** [comment]
@netizen-ais a starfish is not a fish as well, so that's not an argument tbh. The only reason - is used as a minus is due to it being the only sign that resembles minus or dash. It is context based.

**Comment by netizen-ais** [comment]
Ok, so why my numeric keyboard has multiplication, division and plus signs and a not a minus keys? Or are they a star, a cross, a path separator and an hyphen?

**Comment by theScottyJam** [comment]
The primary concern is for those who use screen readers? Just want to make sure - is this a hypothetical concern, or are there people in this conversation who use screen readers with code or know people who use screen readers with code, and who have a really negative experience with the "-" character.

It seems to me that "-" might be a very minor concern compared to a number of other issues I imagine would come up from someone trying to use a screen reader on code. I presume a preferable solution would be to update screen-reader technology to be context-aware, so as it reads the code, it can understand if "-" is intended to be a dash or a minus, among other things, that way we don't have to have everyone update all of their code to be more screen-reader friendly by using characters that aren't even on the keyboard, instead, screen readers can just understand what's already there. And maybe that kind of software already exists.

As mentioned before, the "-" is a "HYPHEN-MINUS", meaning it's valid to use it as either a hyphen or a minus, so using "-" as a minus in a programming language doesn't seem bad to me. Using "equals greater than" to mean "function" seems like a much worse crime, if we're carrying about the accuracy of these characters :).

**Comment by mk-pmb** [comment]
> Using "equals greater than" to mean "function" seems like a much worse crime,

Indeed, especially when we have U+2394 software-function symbol (⎔) for that.

**Comment by phun-ky** [comment]
> Ok, so why my numeric keyboard has multiplication, division and plus signs and a not a minus keys? Or are they a star, a cross, a path separator and an hyphen?

How is that relevant? I am not trying to be snarky here, and I really hoped there was room for a decent conversation around this.

> The primary concern is for those who use screen readers?

No, the primary concern is semantics, secondary is screen readers (or tied)

> It seems to me that "-" might be a very minor concern

It is a minor concern, but is tc39 only for medium + concerns?

`HYPHEN-MINUS` is context based, I have no problem with that what so ever, and I am not trying to be difficult about this. A human can derive that context quite easily, screen readers cannot (and being dependent on AI is not applicable here).

I agree that this is clearly a "nice to have"/minor/"can live without"-feature, and I am not arguing against that. Is it possible to take a step backwards and see it from another perspective to recognize if this is beneficial for the spec? :)

**Comment by theScottyJam** [comment]
> No, the primary concern is semantics, secondary is screen readers (or tied)

Ok, I'll work with that. So the cost-benefit comparison is:

Cost:

There are now two valid ways to type in a minus sign in JavaScript, which is another thing for people to debate over which standard they want to follow - the easy to type hyphen-minus or the unicode minus.

Benefit:

We are using a character that, semantically, is less ambiguous.

---

Could you expand more on what the concrete benefit is for improving semantics. Improved semantics aren't an end-goal, they're a means to a goal.

For example, the word "orange" is technically ambiguous (am I talking about a fruit or a color). We could find ways in our writing to disambiguate the two by, for example, giving alternative spellings to the two forms. It would improve semantics, so should it be done? Should we petition everyone to start doing this? Only if we have a concrete reason to do it - improve semantics, by itself,  isn't that.

**Comment by linusg** [comment]
As Jordan pointed out 8 years ago, this is not the right place for proposal bikeshedding.

> To make a feature request, document the problem and a sketch of the solution with others in the community, including TC39 members. One place to do this is the [TC39 Discourse](https://es.discourse.group/); another is the [Matrix chat room](https://github.com/tc39/how-we-work/blob/HEAD/matrix-guide.md).

**Comment by phun-ky** [comment]
tl;dr; I yield.

----

I appreciate the clarification regarding the design constraints and the importance of maintaining consistency in the ECMAScript specification and tooling ecosystem. While my initial concern stemmed from an **accessibility** and **semantic clarity** perspective (particularly for screen reader users), I understand the practical implications and potential ripple effects such a change could introduce.

Given these considerations, I'm happy to rest my case here. I hope this discussion at least sheds light on how certain characters can be interpreted differently by assistive technologies, and perhaps _encourages_ future collaboration between language communities and accessibility tool developers.

Thanks again for engaging with the idea so constructively.

**Comment by mk-pmb** [comment]
> I agree that this is clearly a "nice to have"/minor/"can live without"-feature,

Also one that you can easily implement in your own flavor of JS and then compile down to standard JS with probably just regexp replacements for most every-day code. (You'd need an actual parser if you want 100% accuracy in avoiding false matches inside regexps, block comments, and string literals; but there are ready-made modules for that. Then again, when you create your own flavor of JS, you can just add a rule that forbids confusing code.)

The nice thing about the JS ecosystem is that there are so many variants of JS that all solve their niche for the few enthusiasts that opt-in to that.

**Comment by phun-ky** [comment]
@mk-pmb "No AI needed. A simple parser module will do. npm has multiple." Sorry, that is a bold statement, a parser cannot be context aware of when we humans read minus vs dash. 

There is no purely mechanical way to resolve all ambiguity without some external or predefined notion of context.

Now, let us leave it as is, and focus on better ways to improve the standard :)

**Comment by tabatkins** [comment]
> There is no purely mechanical way to resolve all ambiguity without some external or predefined notion of context.

In code, there is absolutely a purely mechanical way to tell the difference, that's what the parser *does*. JS knows precisely when something is a minus sign (unary) or subtraction sign (binary) and when it's neither of those things (in a regex, for example, where there's still mechanical usages, like in `[a-z]`, and literal usages, like in `[*/+-]`, that a parser can tell apart).
